### PR TITLE
More optimization and static assert checks in DataHeader definitions

### PR DESCRIPTION
**Pull request by matthiasrichter**
- implementing optimized comparison operator avoiding memcmp
  for Descriptor struct
- adding more static asserts for key parameters
- adding simple benchmark

### CODE REVIEW

**Comment by sawenzel** [comment]
This PR breaks the DataHeader unit test on my MAC.

**Comment by matthiasrichter** [comment]
@sawenzel this has been fixed now, I pushed the corrected commits. Some compilation warnings have been fixed in addition.

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 1]
```diff
@@ -158,6 +158,11 @@ template <int N>
 constexpr std::size_t strLength(const char (&str)[N], std::size_t pos = 0) {
   return ((pos >= N || str[pos] == 0) ? 0 : 1 + strLength(str, pos + 1));
 }
+
+template<typename T>
+constexpr T lower(T a, T b) {
+  return a < b? a:b;
+}
```
Comment (sawenzel) [inline DataFormats/Headers/include/Headers/DataHeader.h]: why not `std::min`?

Comment (matthiasrichter) [inline DataFormats/Headers/include/Headers/DataHeader.h]: simply "a momentary lapse of reason"

**Comment by matthiasrichter** [comment]
The branch is now updated, should be ready for merging

**Pull request merged**